datastore/postgres/get.go

```diff
@@ -83,17 +82,28 @@ func (s *MatcherStore) Get(ctx context.Context, records []*claircore.IndexRecord
 	results := make(map[string][]*claircore.Vulnerability)
 	vulnSet := make(map[string]map[string]struct{})
 	for _, rq := range rqs {
-		rows, ok := resCache[rq.query]
+		rid := rq.record.Package.ID
+		vulns, ok := resCache[rq.query]
 		if !ok {
 			return nil, fmt.Errorf("unexpected vulnerability query: %s", rq.query)
 		}
-		if rows == nil {
-			rows, err = res.Query()
-			if err != nil {
-				res.Close()
-				return nil, err
+		if vulns != nil { // We already have results we don't need to go back to the DB.
+			if _, ok := vulnSet[rid]; !ok {
+				vulnSet[rid] = make(map[string]struct{})
 			}
-			resCache[rq.query] = rows
+			for _, v := range vulns {
+				if _, ok := vulnSet[rid][v.ID]; !ok {
+					vulnSet[rid][v.ID] = struct{}{}
+					results[rid] = append(results[rid], v)
+				}
+			}
+			continue
+		}
+		results[rid] = []*claircore.Vulnerability{}
+		rows, err := res.Query()
```

Comment:
`pgx.Rows` should close at the end of `Next`ing:

```
// Rows is the result set returned from *Conn.Query. Rows must be closed before
// the *Conn can be used again. Rows are closed by explicitly calling Close(),
// calling Next() until it returns false, or when a fatal error occurs.
```

Reply:
The docs I found says this:

```
// Next prepares the next row for reading. It returns true if there is another
// row and false if no more rows are available. It automatically closes rows
// when all rows are read.
Next() bool

// Scan reads the values from the current row into dest values positionally.
// dest can include pointers to core types, values implementing the Scanner
// interface, and nil. nil will skip the value entirely. It is an error to
// call Scan without first calling Next() and checking that it returned true.
Scan(dest ...interface{}) error
```

So I get the happy case will be good, but if `Scan` fails, I don't thinks `rows` will automatically close

Reply:
If `Scan` fails, Rows should be implicitly closed. You can see that [here](https://github.com/jackc/pgx/blob/v4.18.3/rows.go#L197) and [here](https://github.com/jackc/pgx/blob/v4.18.3/rows.go#L173) . It is possible to check by creating a Scan error and trying to call `Values()`:

```go
			if err != nil {
				res.Close()
				_, valErr := rows.Values()
				return nil, fmt.Errorf("failed to scan vulnerability: %v and got a Values error %v", err, valErr)
			}
// failed to start scan: matcher error: failed to scan vulnerability: can't scan into dest[25]: unable to assign to *int64 and got a Values error rows is closed
```

Reply:
I'm fine deferring a `rows.Close()`  in an anonymous function so we avert any change in behavior from the library, just pointing out that it hasn't existed in this function up until this point because it looks like the rows are always closed by the library.